rewrite-java/src/main/java/org/openrewrite/java/dataflow/Dataflow.java

```diff
@@ -25,33 +25,32 @@
 import org.openrewrite.java.tree.J;
 
 import java.util.Iterator;
+import java.util.Optional;
 
 @Incubating(since = "7.24.0")
 @RequiredArgsConstructor
 public class Dataflow {
     final Cursor start;
 
-    public <Source extends Expression, Sink extends J> SinkFlow<Source, Sink> findSinks(LocalFlowSpec<Source, Sink> spec) {
-        Iterator<Cursor> cursorPath = start.getPathAsCursors();
-        while (cursorPath.hasNext()) {
-            Cursor cursor = cursorPath.next();
-            Object value = cursor.getValue();
-            if (spec.getSourceType().isAssignableFrom(value.getClass())) {
-                //noinspection unchecked
-                Source source = (Source) value;
-
-                SinkFlow<Source, Sink> flow = new SinkFlow<>(cursor, spec, start);
-                if (spec.isSource(source, start)) {
-                    ForwardFlow.findSinks(flow);
-                }
-
-                if (flow.isNotEmpty()) {
-                    return flow;
-                }
+    public <Source extends Expression, Sink extends J> Optional<SinkFlow<Source, Sink>> findSinks(LocalFlowSpec<Source, Sink> spec) {;
+        Object value = start.getValue();
+        if (spec.getSourceType().isAssignableFrom(value.getClass())) {
+            //noinspection unchecked
+            Source source = (Source) value;
+            if (!spec.isSource(source, start)) {
+                return Optional.empty();
+            }
+
+            SinkFlow<Source, Sink> flow = new SinkFlow<>(start, spec);
+
+            ForwardFlow.findSinks(flow);
+
+            if (flow.isNotEmpty()) {
+                return Optional.of(flow);
             }
```

Comment:
A source is a source if we are currently visiting a source (or the passed cursor is a source), we don't look at parents anymore.